### PR TITLE
Huge fixture to test slown down

### PR DESCRIPTION
**Pull request by RDruon**
Draft PR that includes a big fixture from jobstats to investigate performance slow down.

### CODE REVIEW

**Comment by RDruon** [comment]
Since I am having issuer with perf, here is the sample code I am using:
```
    use std::time::Instant;

    let repeat = 100000;

    let job = r#"
- job_id:          "SLURM_JOB_machine184_74186:0:ma"
  snapshot_time:   1720516680
  read_bytes:      { samples:           0, unit: bytes, min:        0, max:        0, sum:                0, sumsq:                  0 }
  write_bytes:     { samples:          52, unit: bytes, min:     4096, max:   475136, sum:          5468160, sumsq:      1071040692224 }
  read:            { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  write:           { samples:          52, unit: usecs, min:       12, max:    40081, sum:           692342, sumsq:        17432258604 }
  getattr:         { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  setattr:         { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  punch:           { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  sync:            { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  destroy:         { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  create:          { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  statfs:          { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  get_info:        { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  set_info:        { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  quotactl:        { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }
  prealloc:        { samples:           0, unit: usecs, min:        0, max:        0, sum:                0, sumsq:                  0 }"#;
    let lctl_output = r#"obdfilter.ds002-OST0000.job_stats=
job_stats:"#;

    let input = format!("{lctl_output}{}", job.to_string().repeat(repeat));

    let now = Instant::now();

    let lctl_record = parse_lctl_output(input.as_bytes())?;

    let elapsed = now.elapsed();
    println!("'parse_lctl_output' took: {:.2?}", elapsed);

    
    let input = format!("job_stats:{}", job.to_string().repeat(repeat));
    
    let now = Instant::now();

    serde_yaml::from_str::<JobStatsOst>(&input).unwrap();

    let elapsed = now.elapsed();
    println!("'serde_yaml' took: {:.2?}", elapsed);
```

Here is the output:
```
'parse_lctl_output' took: 38.14s
'serde_yaml' took: 11.31s
```

**Comment by RDruon** [comment]
Caching the jobstats gives us a very low response time:
```
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
3393045

real    0m1.232s
user    0m0.526s
sys     0m0.621s
```
The memory stays stable at ~2.3Gwith peaks at ~6GB when processing jobstats:
```
[root@ds86 ~]# systemctl status lustrefs_exporter | grep Memory
   Memory: 5.1G
[root@ds86 ~]# systemctl status lustrefs_exporter | grep Memory
   Memory: 2.0G
```

**Comment by RDruon** [comment]
Using latest commit, I was able to divide the parsing time by ~3:
```
'parse_lctl_output' took: 23.93s
'parse_lctl_output_old' took: 75.41s
```

It relies on zero-copy parser from combine, https://docs.rs/combine/latest/combine/parser/range/index.html

The memory usage seems to be the same but I will need to test the previous implementation with the same script:
```
[root@node ~]# for i in $(seq 0 1000); do systemctl status lustrefs_exporter.service |grep Memory >> /tmp/test_memory; curl -s http://localhost:32221/metrics -o /dev/null; sleep 1; done; cat /tmp/test_memory | cut -d':' -f2 | numfmt --from=iec | awk -v OFMT='%.10f' '{if(min==""){min=max=$1}; if($1>max) {max=$1}; if($1<min) {min=$1}; total+=$1; count+=1} END {print total/count, max, min}' | xargs -n1 numfmt --to=iec | awk '{if (NR==1) avg=$1; else if (NR==2) max=$1; else if (NR==3) min=$1} END {print "avg " avg " | max " max " | min " min}'
avg 3.7G | max 7.5G | min 1.4G
```

Please find attached the [samply](https://github.com/mstange/samply) profile from latest commit. 
[profile.zip](https://github.com/user-attachments/files/16279199/profile.zip)
You can use https://profiler.firefox.com/ to visualize it

From the flamegraph, it seems most of the time is now spent deserializing the YAML blob.
image

**Comment by RDruon** [comment]
Sub 10s (on my test system) with the latest commit:
```
'read_to_string' took: 222.52ms
'parse_lctl_output' took: 8.96s
```

**Comment by jgrund** [comment]
> Sub 10s (on my test system) with the latest commit:
> 
> ```
> 'read_to_string' took: 222.52ms
> 'parse_lctl_output' took: 8.96s
> ```
> 
> On a real system
> 
> ```
> [root@ds86 ~]# time ./lustre_collector > /dev/null
> 
> real    0m5.466s
> user    0m2.464s
> sys     0m2.950s
> ```

Good improvement. Is this compiled in release mode?

**Comment by breuhan** [comment]
Yes, actually it has to be compiled in release mode in order to work.  There is a slight problem in Debug mode. The memory consumption is down to 440 to 650MB which comes down to reading the source file into memory. 
There is still a spike of 2GB when using the lustrefs_exporter at the end. Due to the conversion into Prometheus string.

**Comment by jgrund** [comment]
> The memory consumption is down to 440 to 650MB which comes down to reading the source file into memory.

Which source file?

> There is still a spike of 2GB when using the lustrefs_exporter at the end. Due to the conversion into Prometheus string.

`PrometheusMetric::render()` ?

**Comment by breuhan** [comment]
> Which source file?
In case of our fixture the read_to_string() function which reads the ds86.txt

> PrometheusMetric::render() ?

Exactly

**Comment by jgrund** [comment]
> Exactly, Raphael thought about using the official prometheus crate which supports writing in a stream.

Link to method for consideration?

**Comment by breuhan** [comment]
> Link to method for consideration?

https://docs.rs/prometheus-client/latest/prometheus_client/encoding/text/fn.encode.html

**Comment by RDruon** [comment]
Caching of jobstats metrics is mandatory:

- To avoid timing out
- To fix the multi consumer use case, in this case, the parsing of jobstats will be done for each scrape, we want to avoid duplicating this information, which will increase the CPU and memory consumption so we need to cache it.

**Comment by RDruon** [comment]
Here are the results of the latest commit running on a live system.
# With jobstats enabled
```
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
0

real    0m0.019s
user    0m0.007s
sys     0m0.002s
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
0

real    0m0.021s
user    0m0.006s
sys     0m0.002s
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
1652586

real    0m0.419s
user    0m0.193s
sys     0m0.297s
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
1652586

real    0m0.451s
user    0m0.215s
sys     0m0.316s
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
1652586

real    0m0.410s
user    0m0.189s
sys     0m0.290s
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
1652586

real    0m0.430s
user    0m0.215s
sys     0m0.276s
[root@ds86 ~]# for i in $(seq 0 100); do systemctl status lustrefs_exporter.service |grep Memory >> /tmp/test_memory; curl -s [http://localhost:32221/metrics -o /dev/null](http://localhost:32221/metrics%20-o%20/dev/null); sleep 1; done; cat /tmp/test_memory |
cut -d':' -f2 | numfmt --from=iec | awk -v OFMT='%.10f' '{if(min==""){min=max=$1}; if($1>max) {max=$1}; if($1<min) {min=$1}; total+=$1; count+=1} END {print total/count, max, min}' | xargs -n1 numfmt --to=iec
| awk '{if (NR==1) avg=$1; else if (NR==2) max=$1; else if (NR==3) min=$1} END {print "avg " avg " | max " max " | min " min}'
avg 1.8G | max 2.5G | min 753M
```

After some time (~10s) the jobstats cache is warm and the exporter start outputting jobstats metrics.
The average response time is ~0.5s.
The average memory consumption is ~2GB, with a max of 2.5GB and a min of 800MB.

# With jobstats disabled
```
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
0

real    0m0.019s
user    0m0.005s
sys     0m0.004s
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
0

real    0m0.021s
user    0m0.005s
sys     0m0.004s
[root@ds86 ~]# time curl -s http://localhost:32221/metrics | grep job | wc -l
0

real    0m0.019s
user    0m0.004s
sys     0m0.004s

[root@ds86 ~]# for i in $(seq 0 100); do systemctl status lustrefs_exporter.service |grep Memory >> /tmp/test_memory; curl -s [http://localhost:32221/metrics -o /dev/null](http://localhost:32221/metrics%20-o%20/dev/null); sleep 1; done; cat /tmp/test_memory | cut -d':' -f2 | numfmt --from=iec | awk -v OFMT='%.10f' '{if(min==""){min=max=$1}; if($1>max) {max=$1}; if($1<min) {min=$1}; total+=$1; count+=1} END {print total/count, max, min}' | xargs -n1 numfmt --to=iec | awk '{if (NR==1) avg=$1; else if (NR==2) max=$1; else if (NR==3) min=$1} END {print "avg " avg " | max " max " | min " min}'
avg 4.1M | max 5.0M | min 3.9M
```
The average response time is ~0.02s.
The average memory consumption is ~4MB, with a max of 5MB and a min of 3.9MB.

File: lustre-collector/src/main.rs [hunk 1]
```diff
@@ -113,36 +112,17 @@ fn run() -> Result<(), LustreCollectorError> {
...
-            Ok(lnetctl_stats_record)
-        });
-
-    let recovery_status_handle =
-        thread::spawn(move || -> Result<Vec<Record>, LustreCollectorError> {
-            let recovery_status_output = get_recovery_status_output()?;
-            let recovery_statuses = parse_recovery_status_output(&recovery_status_output)?;
-
-            Ok(recovery_statuses)
-        });
```
Comment (RDruon) [inline lustre-collector/src/main.rs:145]: Removed `thread::spawn` as the results is now referencing a variable local to the thread.
I tried to use `Arc` without success. 

@jgrund any idea?

**Comment by jgrund** [comment]
No longer relevant.

**Pull request closed without merging**